elkscmd/basic/host-ibmpc.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Architecture Specific routines for CGA
+ * Sep 2024 Takahiro Yamada
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#if __ia16__
+#include <arch/io.h>
+#else
+#define inb(port)           0
+#define inw(port)           0
+#define outb(value,port)
+#define outw(value,port)
+#endif
+#include "host.h"
+#include "basic.h"
+
+#define VIDEO_05_G320x200   0x0005
+#define VIDEO_03_T80x25     0x0003
+
+extern FILE *outfile;
+
+static int gmode = 0;
+static int exit_on = 0;
+
+typedef struct {
+    int x;
+    int y;
+    int fgc;
+    int bgc;
+    int r;
+} xyc_t;
+
+static xyc_t gxyc = {0, 0, 7, 0, 1};
+
+#if __ia16__
+void int_10(unsigned int ax, unsigned int bx,
+            unsigned int cx, unsigned int dx)
+{
+    __asm__ volatile ("push %ds;"
+                      "push %es;"
+                      "push %bp;"
+                      "push %si;"
+                      "push %di;");
+    __asm__ volatile ("int $0x10;"
+                      :
+                      :"a" (ax), "b" (bx), "c" (cx), "d" (dx)
+                      :"memory", "cc");
+    __asm__ volatile ("pop %di;"
+                      "pop %si;"
+                      "pop %bp;"
+                      "pop %es;"
+                      "pop %ds;");
+}
+
+void memclrw(unsigned int offset, unsigned int seg, unsigned int count)
+{
+    __asm__ volatile ("push %ds;"
+                      "push %es;"
+                      "push %bp;"
+                      "push %si;"
+                      "push %di;");
+    __asm__ volatile ("mov %%ax,%%di;"
+                      "mov %%bx,%%es;"
+                      "xor %%ax,%%ax;"
+                      "cld;"
+                      "rep;"
+                      "stosw;"
+                      :
+                      :"a" (offset), "b" (seg), "c" (count)
```

Comment:
It is better to inform the compiler of the exact registers needed (in this case DI), than to tell the compiler to load them into a general register (AX in this case) and then write ASM that the compiler doesn't know about that loads DI. I think you can use the "D" instead of "a" to load DI directly; try it and look at the disassembled output to be sure.

Also, this routine saves SI for no reason - see above for discussion. If possible we should let the compiler generate the function prologue and epilogue.

Using the GCC ASM constraints mechanism is very tricky, non-portable, and generally I can't recommend it. What ends up happening, which is already happening, is that code written only supports a single compiler and then has to be completely rewritten to support OpenWatcom. I think a better approach would be to move all C-based ASM outside the C file and into a .S or .asm file. I realize that we have then the stupid/big problem of two different ASM formats, but I think some discussion might be useful as to coming up with a way to allow simultaneous support of multiple compilers with a single .S ASM file. 

Unfortunately, doing anything like my suggestion right away would take too much time, so we'll hold on that. But I do like the idea of moving all the ASM code out into separate .S files. They could then be added to the C library and used for other applications, rather than only this one. What do you think of that idea?

If the routine(s) is moved out to a .S file, memclrw should be renamed fmemsetw to be more inline with other C library function names.

Reply:
Actually PC-98 uses intC5-pc98.S already,
so if you are ok I will make separate .S file.

Reply:
Do you have preferred file name for the .S file?
I think both int_10 and memclrw would be moved out.

Reply:
> Actually PC-98 uses intC5-pc98.S already,
so if you are ok I will make separate .S file.

Oh, now I see the pc98.S file. I also notice that host-pc98.c has lots of ASM code (also using non-standard prologues etc). But I probably was thinking that PC-98 is special case, which IBM PC is not so much. But ultimately that file is also non-portable, and we're mixing .S and .c files for ASM in the same project for the same platform! Oh well, I suppose that's fine since its' already done.

> Do you have preferred file name for the .S file?

I was thinking just asm-ibmpc.S, and then rename intcC5-pc98.S to asm-pc98.S.

Reply:
In general, I think the best way to handle routines that require ASM is to put them in a .S file instead of using nonstandard compiler ASM extension. That also allows the same routines (graphics in this case) to be used again by yourself in other code, or be put in the C library. It also gets one to think about exactly what the parameters should be in a more general fashion, and also to write specific special prologues for very high speed, rather than using or competing against the C compiler generate prologues.

For examples of very fast prologues, see the kernel library function elks/arch/i86/lib/fmemory.S::fmemsetw. Notice how the function prologue is extremely fast by not pushing onto the stack, and that the parameters to the function are such that a LDS or LES instruction can be used to grab both words of a far pointer. Not every ASM function should be written this way, but memory copy routines are expected to run very quickly. For slower routines, using a gcc-compatible function prologue will help because of our symbolic stack tracing (which I hope to make available in an easier fashion sometime soon).

Since we already have a working fmemsetw for the kernel, perhaps just that function should be copied out of that library fmemory.S file and reused for this work, rather than writing another routine? If all the functions were pulled out into separate .S files, then they could more easily be put into the C library under libc/asm. But we will have to wait a bit for that because more work is required to get them to work in medium model (and regparmcall as well). This shows the extended work required to "do it right", but at least we'd be moving in that direction.